### PR TITLE
deploy: new package for self-deploying Go binaries

### PR DESCRIPTION
**Pull request by rogpeppe**
This allows a Go program to deploy itself
with no other assets required.

### CODE REVIEW

File: deploy/deploy.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
+package deploy
+
+import (
+	"github.com/juju/gocharm/hook"
+	errgo "gopkg.in/errgo.v1"
+	"os"
+	"os/exec"
+)
+
+func DeployCharm(r *hook.Registry, charmDir string, appName string) error {
```
Comment (alesstimec) [inline deploy/deploy.go]: godoc

File: deploy/main.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+)
+
+var (
+	deployFlag     string
+	buildFlag      string
+	runHookFlag    string
+	noCompressFlag bool
+)
+
+func MainFlags() {
```
Comment (alesstimec) [inline deploy/main.go]: goes this need to be exported?

File: deploy/main.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
+// Package deploy implements functionality for
+// building and deploying Go charms.
+//
+// func main() {
```
Comment (alesstimec) [inline deploy/main.go]: ?

File: deploy/main.go (hunk 2 above)
Comment (rogpeppe) [inline deploy/main.go]: Yeah, I've added a doc comment.

File: deploy/main.go (hunk 3 above)
Comment (rogpeppe) [inline deploy/main.go]: Fixed, and added an example.

File: deploy/deploy.go (hunk 1 above)
Comment (rogpeppe) [inline deploy/deploy.go]: I don't think it warranted export so I unexported it.

**Pull request merged**